dataGenerator.js

```diff
@@ -23,149 +23,939 @@ var db = require('./server/db.js');
+                                if (err) {
+                                    return callback(err, null);
+                                }
+
+                                commentActionsAdded++;
+                                if (commentActionsAdded === 2 * commentActionsPerQuestion * actualCommentsAdded) {
+                                    createReplyActions();
+                                }
+                            });
+
+                            let reply;
+
+                            if (Math.floor(Math.random() * 100) > (100 - mentionsPercentage)) {
+                                reply = datagenInfo.comment + ' @' + userToMention;
+                            } else {
+                                reply = datagenInfo.comment;
+                            }
+
+                            questions.addReply(question.comments[j]._id, userId, reply, function (err, res) {
+                                if (err) {
+                                    return callback(err, null);
+                                } else {
+                                    actualRepliesAdded++
+                                }
+
+                                commentActionsAdded++;
+
+                                if (commentActionsAdded === 2 * commentActionsPerQuestion * actualCommentsAdded) {
+                                    createReplyActions();
+                                }
+                            });
+                        } else {
+                            commentActionsAdded += 2;
+
+                            if (commentActionsAdded === 2 * commentActionsPerQuestion * actualCommentsAdded) {
+                                createReplyActions();
+                            }
+                        }
+
+                    }
+                }
+            });
+        }
+    });
+}
+
+/**
+ * This function creates reply actions, such as likes, and dislikes.
+ * It uses the globally defined settings to do so
+ */
+var createReplyActions = function () {
+    for (var id = 1; id <= totalNumberOfQuestions; id++) {
+        addReplyActions(id, function (err, res) { });
+    }
+}
+
+/**
+ * This function adds the action replies to the question specified
+ * 
+ * @param {integer} questionId 
+ * @param {function} callback 
+ */
+var addReplyActions = function (questionId, callback) {
+    db.lookupQuestion({ id: questionId }, function (err, question) {
+        questionId = question._id;
+        if (err) {
+            return callback(err, null);
+        } else if (!question) {
+            return callback('Could not find the question', null);
+        } else {
+            var answeredList = [];
+
+            for (var i in question.correctAttempts) {
+                answeredList.push(question.correctAttempts[i].id);
+            }
+
+            for (var j = 0; j < question.comments.length; j++) {
```

Comment:
this is so efficient. Can we use some dynamic programming to optimise this?

Reply:
nopenopenopenope